### ISSUE TITLE
deleting images on overlayfs is very slow

### ISSUE THREAD

**Issue by vikstrous** [docker/docker#17551]
In both docker 1.8.3 and 1.9.0-rc3 deleting images on overlayfs is very slow.

For example, I had 600 images. I tried to delete an image with 14 layers. It took 6 seconds on 1.8.3 and 5 seconds on 1.9.0-rc3.

**Comment by thaJeztah** [comment]
@vikstrous is this a regression since 1.8.2? Have you been able to compare this with a pre 1.8.3 version?

**Comment by cpuguy83** [comment]
Honestly, I think this is likely the nature of overlayfs. It creates tons and tons of files.
I literally did an `rm -rf` on my (old) overlay graph dir and it took around 15min to delete it all.

**Comment by vikstrous** [comment]
I have not tried it on other versions. I agree that this is probably a limitation of overlayfs. I don't know the details of overlayfs and how it's used by docker. I'll try to read up on that more. Do you guys know if there's anything that can be done about it? Anything in docker or the kernel?

**Comment by thaJeztah** [comment]
@vikstrous you're in luck; we just merged the initial version of our "storage driver guide"; https://github.com/docker/docker/blob/master/docs/userguide/storagedriver/index.md

This page details OverlayFS; https://github.com/docker/docker/blob/master/docs/userguide/storagedriver/overlayfs-driver.md

**Comment by vikstrous** [comment]
@cpuguy83 I thought that there are hardlinks because overlayfs doesn't support more than two layers. Am I reading this wrong?

https://www.kernel.org/doc/Documentation/filesystems/overlayfs.txt
```
Multiple lower layers
---------------------

Multiple lower layers can now be given using the the colon (":") as a
separator character between the directory names.  For example:

  mount -t overlay overlay -olowerdir=/lower1:/lower2:/lower3 /merged

As the example shows, "upperdir=" and "workdir=" may be omitted.  In
that case the overlay will be read-only.

The specified lower directories will be stacked beginning from the
rightmost one and going left.  In the above example lower1 will be the
top, lower2 the middle and lower3 the bottom layer.
```

**Comment by vikstrous** [comment]
I confirmed locally that multiple lower directories seem to work (I tried only simple files).

@alexlarsson Was there a good reason to not use multiple lower directories in the overlay driver? I'm interested it trying to re-write it use the multiple lower directories feature.

**Comment by vikstrous** [comment]
Update for anyone who stumbles upon this issue: I tried writing a better overlayfs driver and failed due to bugs and limitations in overlayfs https://github.com/docker/docker/pull/18560
This issue is basically blocked on overlayfs getting better.

**Comment by thaJeztah** [comment]
Yes we really should try getting the attention of the overlay maintainer upstream; I know @LK4D4 tried to (he submitted a patch to fix the issue with sockets), but no response whatsoever. :cry:

**Comment by LK4D4** [comment]
@dmcgowan wrote new driver and all patches were merged into kernel. I think we can close this for now.
